### ISSUE TITLE
Injecting with destination local IP address

### ISSUE THREAD

**Issue by edlau2** [basil00/Divert#39]
Is it possible using WinDivert to inject a modified packet by changing the destination IP to a valid local interface address? My goal is to intercept packets arriving from another PC and divert them to a local proxy server application. I can capture the packets from PC A, destined for PC C, on PC B, in the middle, and modify the packet accordingly. WinDivertSendEx(), return successfully, and the completion function does not indicate any errors. This is using a handle with WINDIVERT_LAYER_NETWORK_FORWARD specified. However, I see not indication that the packet ever makes it anywhere.

According to the MSFT 's "Packet Injection Functions" documentation, at https://msdn.microsoft.com/en-us/library/windows/hardware/ff569975(v=vs.85).aspx, "Cross-layer injection is enabled if the callout can supply all needed information that is required by the injection function, and the net buffer list has the format expected by the injection function. For example, a callout can capture a packet at the forward path, modify its destination address to that of the local computer, and call FwpsInjectTransportReceiveAsync0 to redirect the packet into the local computer's TCP/IP stack."

This makes me believe it is possible. I notice the WinDivert driver uses "FwpsInjectForwardAsync0" in the windivert_write() function, I did attempt to change that to use "FwpsInjectTransportReceiveAsync0" as an experiment (with the same results as the unmodified function). In both cases, I have tried WINDIVERT_DIRECTION_INBOUND and outbound, and uses the IfIdx and IfSubIdx of the adapter with the IP address I have modified the packet's destination address too. To verify that packet modification was working as planned, I modified the source address of the packet  instead to that of the local NIC and it injected and sent successfully. By successfully I mean I could view it using Wireshark and verify the checksums.

Based on the quote from the documentation, maybe I am missing something regarding this part: " if the callout can supply all needed information that is required by the injection function, and the net buffer list has the format expected by the injection function". 

Does anyone know if this can be done and if so, how? Thanks in advance

**Comment by basil00** [comment]
The way to do this is to open two WinDivert handles.  The first (with layer=NETWORK_FORWARD) is to intercept packets from the forward layer, and the second (with layer=NETWORK) is to inject the redirected packets towards the local machine.  Also the FwpsInjectTransport... function is for the transport layer, whereas WinDivert works at the network layer.

**Comment by edlau2** [comment]
Thanks basil, I had tried that before, and didn't think it had worked - it must have been a late night. My code to do just that was still in place, I uncommented it out and lo and behold it works. Next step is to get the correct filters in place for the return path, they are already constructed but using the forward layer so I'll have to change that I assume - it all makes sense now in hindsight. Again, thanks for the clarifications, I admit I haven't thoroughly read all pertinent MSDN docs re: WFP
